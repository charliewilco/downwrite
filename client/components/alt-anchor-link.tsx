import * as React from "react";
import classNames from "../utils/classnames";

interface AltAnchorProps<T> extends React.AnchorHTMLAttributes<T> {
  space?: string;
}

export default function AltAnchorLink({
  space,
  className,
  ...props
}: AltAnchorProps<any>): JSX.Element {
  const cx = classNames("AltLink", className);
  return <a {...props} className={cx} />;
}
