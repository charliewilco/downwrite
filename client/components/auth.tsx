import * as React from "react";
import Cookies from "universal-cookie";
import Router from "next/router";
import * as jwt from "jwt-decode";
import addDays from "date-fns/add_days";
import { NextComponentClass } from "next";
import { __IS_TEST__ } from "../utils/dev";

// NOTE:
// This component should passdown the state of authed from withAuthCheck() HOC
// There is only one single point of state that needs to be rendered

// One other pattern we could consider is passing down user and token as state
// and login and logout functions from authed. this would allow an initial check
// of the cookie on a refresh and as the user is logged in have an updated source of the token
// this would solve that single point of state to be updated.
// We would pass down signIn and signOut to <Login /> & <Register />

/*
	<Auth>
		{(authed, login, logout, token, user) => <App {...args} />}
	</Auth>
*/

// state needs to evaluate the existence of
// token + decode the content of the token

const cookie = new Cookies();
const COOKIE_EXPIRATION = 180;
const cookieOptions = {
  path: "/",
  expires: addDays(Date.now(), COOKIE_EXPIRATION)
};

interface IAuthProps {
  children: React.ReactNode;
  token: string;
  authed: boolean;
}

interface IAuthState {
  token: string;
  name?: string;
  authed: boolean;
}

export interface IAuthActions {
  signIn: (authed: boolean, token: string) => void;
  signOut: () => void;
}

export interface IAuthContext extends IAuthState, IAuthActions {}

interface IToken {
  name: string;
  user: string;
}

const EMPTY_USER: IToken = {
  user: null,
  name: null
};

export const AuthContext = React.createContext({} as IAuthContext);

// Should be able to just request user details from another call
export default class AuthMegaProvider extends React.Component<
  IAuthProps,
  IAuthState
> {
  constructor(props: IAuthProps) {
    super(props);

    let token = this.props.token || cookie.get("DW_TOKEN");

    let __TOKEN_EXISTS__: boolean = token !== undefined && token !== "undefined";
    const { name } = __TOKEN_EXISTS__ ? jwt<IToken>(token) : EMPTY_USER;

    this.state = {
      token,
      authed: __TOKEN_EXISTS__,
      name: name || null
    };
  }

  private signIn = (authed: boolean, token: string): void => {
    const { name } = jwt(token);
    this.setState({ authed, token, name });
    cookie.set("DW_TOKEN", token, cookieOptions);
  };

  private signOut = (): void => {
    this.setState({
      authed: false,
      token: undefined,
      name: undefined
    });
  };

  public componentDidUpdate(prevProps: IAuthProps, prevState: IAuthState): void {
    const { authed } = this.state;
    if (prevState.authed !== authed) {
      Router.push({ pathname: authed ? "/" : "/login" });

      if (!authed) {
        cookie.remove("DW_TOKEN", cookieOptions);
      }
    }
  }

  public render(): JSX.Element {
    return (
      <AuthContext.Provider
        value={{
          ...this.state,
          signIn: this.signIn,
          signOut: this.signOut
        }}>
        {this.props.children}
      </AuthContext.Provider>
    );
  }
}

export const AuthConsumer = AuthContext.Consumer;

// TODO: Remove
// Component should be NextStatelessComponent type
export const withAuth = (Component: NextComponentClass<any>) => {
  return class extends React.Component<any, any> {
    public static displayName = `withAuth(${Component.displayName ||
      Component.name})`;

    public static getInitialProps = Component.getInitialProps;

    public render(): JSX.Element {
      return (
        <AuthContext.Consumer>
          {(auth: IAuthActions) => <Component {...this.props} {...auth} />}
        </AuthContext.Consumer>
      );
    }
  };
};
