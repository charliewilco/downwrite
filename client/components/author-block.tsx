import * as React from "react";
import Link from "next/link";
import Avatar from "./avatar";

interface IAuthorProps {
  authed: boolean;
  name: string;
  colors: string[];
}

function TeaserCopy(): JSX.Element {
  return <>You can write and share on Downwrite, you can sign up or log in </>;
}

export default function Author(props: IAuthorProps): JSX.Element {
  return (
    <aside className="AuthorBlock">
      <header className="AuthorBlockHeader">
        <Avatar colors={props.colors} />
        <h6 className="AuthorBlockTitle">Post from {props.name}</h6>
      </header>
      {!props.authed && (
        <>
          <hr />
          <p className="AuthorBlockContent">
            <TeaserCopy />
            <Link prefetch href="/login">
              <a>here</a>
            </Link>
          </p>
        </>
      )}
    </aside>
  );
}
