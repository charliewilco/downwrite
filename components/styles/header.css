.AppHeader {
  display: flex;
  background: var(--background);
  align-items: center;
  justify-content: space-between;
  padding-top: 16px;
  padding-bottom: 16px;
  padding-left: 8px;
  padding-right: 8px;
}

.AppHeaderTitle {
  font-size: 16px;
  font-style: normal;
  font-family: var(--sans);
  line-height: 1;
  font-weight: 700;
}

.AppHeaderLink {
  margin-left: 12px;
  display: block;
  cursor: pointer;
  transition: color 375ms ease-in-out;
  color: var(--headerLogoLink) !important;
}

.AppHeaderLink:hover {
  color: var(--blue700);
}

.AppHeaderNav {
  display: flex;
  align-items: center;
}

@media (min-width: 57.75rem) {
  .AppHeaderTitle {
    font-size: 20px;
  }
}
