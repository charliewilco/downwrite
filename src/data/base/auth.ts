import decode from "jwt-decode";
import base64 from "base-64";
import { BehaviorSubject } from "rxjs";

import { APIClient } from "@data/client";
import type { IAppState } from "@data/types";
import { TOKEN_NAME } from "@shared/constants";
import type { TokenContents } from "@shared/types";

export interface ICurrentUserState {
  username?: string;
  id?: string;
  authed: boolean;
}

export interface IRegisterValues {
  username: string;
  password: string;
  legalChecked: boolean;
  email: string;
}

export interface ILoginValues {
  user: string;
  password: string;
}

export class Auth {
  #internalState: ICurrentUserState = {
    username: "",
    id: "",
    authed: false
  };
  readonly state = new BehaviorSubject<ICurrentUserState>({
    username: "",
    id: "",
    authed: false
  });
  #client: APIClient;
  #store: IAppState;
  constructor(_graphql: APIClient, store: IAppState) {
    this.#client = _graphql;
    this.#store = store;
  }

  async check() {
    const cookies = this.#client.cookies.getAll();

    if (cookies[TOKEN_NAME]) {
      try {
        const value = await this.#client.isMe();

        if (value) {
          console.log(value);
          this.#internalState.id = value.me.details.id;
          this.#internalState.username = value.me.details.username;
          this.#callNext();
        }
      } catch (error) {
        this.#internalState.id = "";
        this.#internalState.username = "";

        console.log(error.message);
      }
    }
  }

  #callNext() {
    this.state.next({
      ...this.#internalState,
      authed: !!this.#internalState.username && !!this.#internalState.id
    });
  }

  onLogin(username: string, id: string) {
    this.#internalState.username = username;
    this.#internalState.id = id;

    this.#callNext();
  }

  onLogout() {
    this.#internalState.username = "";
    this.#internalState.id = "";

    this.#callNext();
  }

  async login(values: ILoginValues) {
    try {
      const value = await this.#client.loginUser({
        username: values.user,
        password: base64.encode(values.password)
      });

      const token = value.authenticateUser.token;

      const d = decode<TokenContents>(token);
      this.onLogin(d.name, d.user);
      this.#client.setToken(token);

      for (const n of this.#store.notifications.getAll()) {
        this.#store.notifications.remove(n.id);
      }
    } catch (error) {
      this.#store.notifications.error(error.message);
    }
  }

  async register({ legalChecked, ...values }: IRegisterValues) {
    if (legalChecked) {
      try {
        const value = await this.#client.createUser({
          ...values,
          password: base64.encode(values.password)
        });

        if (value.createUser.token) {
          const token = value.createUser.token;

          const d = decode<TokenContents>(token);
          this.onLogin(d.name, d.user);
          this.#client.setToken(token);

          for (const n of this.#store.notifications.getAll()) {
            this.#store.notifications.remove(n.id);
          }
        }
      } catch (error) {
        this.#store.notifications.error(error.message);
      }
    }
  }
}
