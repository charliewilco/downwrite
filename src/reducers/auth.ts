export enum AuthActions {
  SIGN_IN = "SIGN_IN",
  SIGN_OUT = "SIGN_OUT"
}

export interface IAuthState {
  token: string;
  name?: string;
  authed: boolean;
}

export interface IAuthReducerAction {
  type: AuthActions;
  payload?: IAuthState;
}

export function reducer(state: IAuthState, action: IAuthReducerAction) {
  switch (action.type) {
    case AuthActions.SIGN_IN:
      return { ...action.payload };
    case AuthActions.SIGN_OUT:
      return { token: null, authed: false, name: null };
    default:
      throw new Error();
  }
}
