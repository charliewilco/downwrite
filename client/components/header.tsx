import * as React from "react";
import styled from "styled-components";
import Link from "next/link";
import Logo from "./logo";
import AltAnchor from "./alt-anchor-link";
import { NavIcon } from "./icons";
import * as DefaultStyles from "../utils/defaultStyles";
import { withAuth } from "./auth";

const MenuContainer = styled.nav`
  display: flex;
  align-items: center;
`;

const HomeLink = styled.a`
  margin-left: 12px;
  display: block;
  cursor: pointer;
  transition: color 375ms ease-in-out;
  font-weight: 700;
  color: ${props => props.theme.headerLogoLink} !important;
  &:hover {
    color: ${DefaultStyles.colors.blue700};
  }
`;

const ToggleButton = styled.button`
  appearance: none;
  outline: none;
  border: 0;
  font-family: inherit;
  background: none;
  box-sizing: inherit;
`;

const HeaderTitle = styled.h1`
  font-size: 16px;
  font-style: normal;
  font-family: ${DefaultStyles.fonts.sans};
  line-height: 1;
  font-weight: 400;

  @media (min-width: 57.75rem) {
    font-size: 20px;
  }
`;

const Header = styled.header`
  display: flex;
  background: ${props => props.theme.background};
  align-items: center;
  justify-content: space-between;
  padding-top: 16px;
  padding-bottom: 16px;
  padding-left: 8px;
  padding-right: 8px;

  .Navicon #NaviconGroup {
    fill: ${props => props.theme.color};
  }
`;

interface IHeaderProps {
  authed: boolean;
  onClick: () => void;
  route: string;
}

class UIHeader extends React.Component<IHeaderProps, {}> {
  render() {
    const { authed, onClick, route } = this.props;

    const url = !authed
      ? {
          href: "/login",
          as: "/"
        }
      : {
          href: "/",
          as: "/"
        };

    return !(route === "/login") ? (
      <Header data-testid="APP_HEADER">
        <MenuContainer>
          <Logo />
          <HeaderTitle data-testid="APP_HEADER_TITLE">
            <Link {...url}>
              <HomeLink>Downwrite</HomeLink>
            </Link>
          </HeaderTitle>
        </MenuContainer>
        {authed ? (
          <MenuContainer>
            <Link prefetch href="/new" as="/new">
              <AltAnchor space="right">New</AltAnchor>
            </Link>
            <ToggleButton onClick={onClick}>
              <NavIcon className="Navicon" />
            </ToggleButton>
          </MenuContainer>
        ) : (
          <MenuContainer>
            <Link prefetch href="/login">
              <AltAnchor space="right">Login or Sign Up</AltAnchor>
            </Link>
          </MenuContainer>
        )}
      </Header>
    ) : null;
  }
}

export default withAuth(UIHeader);
