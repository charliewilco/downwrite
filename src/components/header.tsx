import * as React from "react";
import Link from "next/link";
import { useRouter } from "next/router";
import Logo from "./logo";
import AltAnchor from "./alt-anchor-link";
import { AuthContext, AuthContextType } from "./auth";
import DropdownUI from "./dropdown-ui";

export function UIHeader(): JSX.Element {
  const [{ authed }] = React.useContext<AuthContextType>(AuthContext);
  const router = useRouter();

  const style: React.CSSProperties = { marginRight: 16 };

  const isLogin: boolean = router ? router.route === "/login" : false;

  return !isLogin ? (
    <header className="AppHeader" data-testid="APP_HEADER">
      <nav className="AppHeaderNav">
        <Logo />
        <h1 className="AppHeaderTitle" data-testid="APP_HEADER_TITLE">
          <Link href={!authed ? "/login" : "/"}>
            <a className="AppHeaderLink">Downwrite</a>
          </Link>
        </h1>
      </nav>
      {authed ? (
        <nav className="AppHeaderNav">
          <Link href="/new">
            <AltAnchor style={style}>New</AltAnchor>
          </Link>
          <DropdownUI />
        </nav>
      ) : (
        <nav className="AppHeaderNav">
          <Link href="/login">
            <AltAnchor style={style}>Login or Sign Up</AltAnchor>
          </Link>
        </nav>
      )}
    </header>
  ) : null;
}

export default UIHeader;
