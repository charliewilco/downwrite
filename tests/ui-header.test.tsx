import * as React from "react";
import * as ReactDOM from "react-dom";
import TestRenderer from "react-test-renderer";

import { UIHeader } from "../components/header";
import { LinkProps } from "next/link";
import { MockAuthProvider } from "../components/auth";
import { act } from "react-dom/test-utils";

jest.mock("universal-cookie", () => {
  return class Cookie {};
});
jest.mock("jwt-decode");

jest.mock("next/link", () => {
  return jest.fn((props: React.PropsWithChildren<LinkProps>) => (
    <>{props.children}</>
  ));
});

const app = (
  <MockAuthProvider>
    <UIHeader />
  </MockAuthProvider>
);

let container: HTMLDivElement;

const testMock = TestRenderer.create(app);

beforeEach(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  document.body.removeChild(container);
  container = null;
});

// NOTE: test broken by upgrading @testing-library
describe("Header Component", () => {
  it("contains application name", () => {
    act(() => {
      ReactDOM.render(app, container);
    });

    const title = container.querySelector("h1");
    const header = container.querySelector("header");

    expect(title.textContent).toBe("Downwrite");
    expect(header).toBeTruthy();
  });

  it("matches snapshot", () => {
    expect(testMock.toJSON()).toMatchSnapshot();
  });
});
