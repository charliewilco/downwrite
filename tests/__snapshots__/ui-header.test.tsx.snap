// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component matches snapshot 1`] = `
<header
  className="AppHeader"
  data-testid="APP_HEADER"
>
  <nav
    className="AppHeaderNav"
  >
    <div
      className="Logo"
    >
      <svg
        className="LogoIcon"
        height={50}
        viewBox="0 0 33 50"
        width={33}
      >
        <title>
          Downwrite Logo
        </title>
        <defs>
          <radialGradient
            fx="50%"
            fy="50%"
            id="a"
            r="143.827%"
          >
            <stop
              offset="0%"
              stopColor="#FFCE00"
            />
            <stop
              offset="16.39%"
              stopColor="#FFD600"
            />
            <stop
              offset="57.9%"
              stopColor="#FFE600"
            />
            <stop
              offset="100%"
              stopColor="#FFCE00"
            />
          </radialGradient>
          <radialGradient
            fx="50%"
            fy="50%"
            id="b"
            r="143.827%"
          >
            <stop
              offset="0%"
              stopColor="#FFB600"
            />
            <stop
              offset="52.49%"
              stopColor="#FFC200"
            />
            <stop
              offset="100%"
              stopColor="#F4A903"
            />
          </radialGradient>
        </defs>
        <g
          fill="none"
          fillRule="evenodd"
        >
          <path
            d="M16.469 5.275V30.97c1.202 0 2.176.965 2.176 2.155s-.974 2.155-2.176 2.155v9.444l10.101-9.594L16.47 5.275z"
            fill="url(#a)"
          />
          <path
            d="M14.293 33.126c0-1.19.974-2.155 2.176-2.155V5.275L6.367 35.13l10.102 9.594v-9.444a2.165 2.165 0 0 1-2.176-2.155z"
            fill="url(#b)"
          />
          <path
            d="M32.938 50H0V0h32.938v50zM1.63 48.47h29.677V1.53H1.63v46.94z"
            fill="#FFC200"
          />
        </g>
      </svg>
    </div>
    <h1
      className="AppHeaderTitle"
      data-testid="APP_HEADER_TITLE"
    >
      <a
        className="AppHeaderLink"
      >
        Downwrite
      </a>
    </h1>
  </nav>
  <nav
    className="AppHeaderNav"
  >
    <a
      className="AltLink"
      style={
        Object {
          "marginRight": 16,
        }
      }
    >
      Login or Sign Up
    </a>
  </nav>
</header>
`;
